src/DotNetWorker.Core/Converters/Converter/DefaultInputConverterProvider.cs

```diff
@@ -0,0 +1,119 @@
+        /// </summary>
+        private readonly ConcurrentDictionary<string, IInputConverter> _converterCache = new();
+
+        /// <summary>
+        /// Stores the default converter instances.
+        /// This is an ordered sub set of what is present in _converterCache.
+        /// </summary>
+        private IReadOnlyList<IInputConverter> _defaultConverters;
+
+        public DefaultInputConverterProvider(IOptions<WorkerOptions> workerOptions, IServiceProvider serviceProvider)
+        {
+            _workerOptions = workerOptions?.Value ?? throw new ArgumentNullException(nameof(workerOptions));
+            _serviceProvider = serviceProvider ?? throw new ArgumentNullException(nameof(serviceProvider));
+            
+            InitializeConverterCacheWithDefaultConverters();
+        }
+                
+        /// <summary>
+        /// Gets an ordered collection of default converter instances.
+        /// </summary>
+        public IEnumerable<IInputConverter> DefaultConverters => _defaultConverters;
+
+        /// <summary>
+        /// Gets an instance of the converter for the type requested.
+        /// </summary>
+        /// <param name="converterTypeName">The assembly qualified name of the type for which we are requesting an IInputConverter instance.</param>
+        /// <exception cref="ArgumentNullException">Throws when the converterTypeName param is null.</exception>
+        /// <returns>IConverter instance of the requested type.</returns>
+        public IInputConverter GetOrCreateConverterInstance(string converterTypeName)
+        {
+            if (converterTypeName is null)
+            {
+                throw new ArgumentNullException((nameof(converterTypeName)));
+            }
+
+            // Get from cache or create the instance and cache
+            return _converterCache.GetOrAdd(converterTypeName, (key, converterTypeAssemblyQualifiedName) =>
+            {
+                // Create the instance and cache that against the assembly qualified name of the type.
+                var converterType = Type.GetType(converterTypeAssemblyQualifiedName);
+
+                if (converterType is null)
+                {
+                    throw new InvalidOperationException($"Could not create an instance of {converterTypeAssemblyQualifiedName}");
+                }
+
+                EnsureTypeCanBeAssigned(converterType);
+
+                return (IInputConverter)ActivatorUtilities.CreateInstance(_serviceProvider, converterType);
+
+            }, converterTypeName);
+        }
+
+        /// <summary>
+        /// Initializes the converter cache from worker options.
+        /// </summary>
+        private void InitializeConverterCacheWithDefaultConverters()
+        {
+            if (_workerOptions.InputConverters is null || _workerOptions.InputConverters.Count == 0)
+            {
+                throw new InvalidOperationException("No input converters found in worker options!");
```

Comment:
nit: if we don't use the `!` in other error messages, we should avoid doing so here :D